lib/valueflow.cpp

```diff
@@ -2654,6 +2654,189 @@ static void valueFlowAfterCondition(TokenList *tokenlist, SymbolDatabase* symbol
     }
 }
 
+static void valueFlowForwardLifetime(Token * tok, TokenList *tokenlist, ErrorLogger *errorLogger, const Settings *settings)
+{
+    const Token * assignTok = tok->astParent();
+    // Assignment
+    if (!assignTok || (assignTok->str() != "=") || assignTok->astParent())
+        return;
+
+    // Lhs should be a variable
+    if (!assignTok->astOperand1() || !assignTok->astOperand1()->varId())
+        return;
+    const Variable *var = assignTok->astOperand1()->variable();
+    if (!var || (!var->isLocal() && !var->isGlobal() && !var->isArgument()))
+        return;
+
+    const Token * const endOfVarScope = var->typeStartToken()->scope()->bodyEnd;
+
+    // Rhs values..
+    if (!assignTok->astOperand2() || assignTok->astOperand2()->values().empty())
+        return;
+
+    std::list<ValueFlow::Value> values = assignTok->astOperand2()->values();
+
+    // Static variable initialisation?
+    if (var->isStatic() && var->nameToken() == assignTok->astOperand1())
+        changeKnownToPossible(values);
+
+    // Skip RHS
+    const Token * nextExpression = nextAfterAstRightmostLeaf(assignTok);
+
+    valueFlowForward(const_cast<Token *>(nextExpression), endOfVarScope, var, var->declarationId(), values, false, false, tokenlist, errorLogger, settings);
+}
+
+static const Variable * getLifetimeVariable(const Token * tok)
+{
+    const Variable * var = tok->variable();
+    if(!var)
+        return nullptr;
+    if(var->isReference() || var->isRValueReference()) {
+        for(const ValueFlow::Value& v:tok->values()) {
+            if(!v.isLifetimeValue() && !v.tokvalue)
+                continue;
+            const Variable * var2 = getLifetimeVariable(v.tokvalue);
+            if(var2)
+                return var2;
+        }
+        return nullptr;
+    }
+    return var;
+}
+
+struct Lambda
+{
+    explicit Lambda(const Token * tok)
+    : capture(nullptr), arguments(nullptr), returnTok(nullptr), bodyTok(nullptr)
+    {
+        if(!Token::simpleMatch(tok, "[") || !tok->link())
+            return;
+        capture = tok;
+
+        if(Token::simpleMatch(capture->link(), "] (")) {
+            arguments = capture->link()->next();
+        }
+        const Token * afterArguments = arguments ? arguments->link()->next() : capture->link()->next();
+        if(Token::simpleMatch(afterArguments, "->")) {
```

Comment:
This condition is never true. The `->` operators are replaced with `.` in the tokenlist. For information, if you need to know if the original code says `->` then you can use `Token::originalName()`.

Reply:
I see, I thought it was only replaced for member accesses.